### PR TITLE
DON-4388 Add search input summary

### PR DESCRIPTION
**Pull request by henrik-sky**
<!--
Thanks for contributing to Backpack :pray:

Please include a description of the changes you are introducing and some screenshots if appropriate.

Please ensure your pull request title is clear as it will be used to generate the changelog.

Add `major`, `minor` or `patch` label depending on the change according to [semver](semver.org) or `skip-changelog` if the change shouldn't be added to the changelog (e.g. a change to a test or documentation)
-->

[Figma link](https://www.figma.com/file/irZ3YBx8vOm16ICkAr7mB3/Backpack-Components?type=design&node-id=42020-91608&mode=design&t=oXDFSys4uajEIqP5-4) 

We're adding a new component to be used in the AppSearchModal that will allow us to either have text or an icon before the search term.
~This has changed the API for AppSearchModal hence I have added the `major` label.~

Additionally I achieved this change by changing BpkTextField to have a Composable input instead of the icon one. I added overloads to keep existing functionality with the icon parameter.

## Default

| Day                                                                                                                                                                                                                                                                 | Night                                                                                                                                                                                                                                                                              |
|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| <img src="https://github.com/Skyscanner/backpack-android/assets/141921186/bbbebac1-7ba4-40fa-860a-b5e988f5b4a2" alt="BpkSearchInputSummary component showing states including icon, text prefix with and without text entered" width="375" /> | <img src="https://github.com/Skyscanner/backpack-android/assets/141921186/5f44b8f1-1211-4af7-a72b-d15849f14270" alt="BpkSearchInputSummary component showing states including icon, text prefix with and without text entered - dark mode" width="375" /> |



+ [ ] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [ ] Component `README.md`
+ [ ] Tests

If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)

### CODE REVIEW

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/searchinputsummary/BpkSearchInputSummary.kt [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                        icon = summary.prefix.icon,
+                        contentDescription = null,
+                        size = BpkIconSize.Large,
+                        modifier = Modifier.padding(start = BpkSpacing.Sm),
+                        tint = animateColorAsState(
+                            when (LocalFieldStatus.current) {
+                                is BpkFieldStatus.Disabled -> BpkTheme.colors.textDisabled
+                                else -> BpkTheme.colors.textSecondary
+                            },
+                        ).value,
```
Comment (henrik-sky) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/searchinputsummary/BpkSearchInputSummary.kt:83]: Copied from BpkTextField

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/textfield/BpkTextField.kt [hunk 2]
```diff
@@ -58,7 +63,55 @@ fun BpkTextField(
         modifier = modifier,
         readOnly = readOnly,
         placeholder = placeholder,
-        icon = icon,
+        prefix = {
```
Comment (henrik-sky) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/textfield/BpkTextField.kt:66]: Open to suggestions for the name of this

File: app/src/main/java/net/skyscanner/backpack/demo/compose/BpkSearchInputSummaryStory.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import net.skyscanner.backpack.compose.textfield.BpkClearAction
+import net.skyscanner.backpack.compose.tokens.Hotels
+import net.skyscanner.backpack.demo.R
+import net.skyscanner.backpack.demo.components.SearchInputSummaryComponent
+import net.skyscanner.backpack.demo.meta.ComposeStory
+
+@Composable
+@SearchInputSummaryComponent
+@ComposeStory("Prefix Text")
+fun PrefixTextSearchInputSummary(modifier: Modifier = Modifier) {
```
Comment (marianeum) [inline app/src/main/java/net/skyscanner/backpack/demo/compose/BpkSearchInputSummaryStory.kt:43]: can you change these to be one story and include a version with and without input to closer match what the figma samples are

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/appsearchmodal/BpkAppSearchModal.kt [hunk 4]
```diff
@@ -95,13 +96,14 @@ fun BpkAppSearchModal(
...
     ) {
-        BpkAppSearchModalImpl(
-            inputText = inputText,
-            inputHint = inputHint,
-            results = results,
-            onInputChanged = onInputChanged,
-            clearAction = clearAction,
-            behaviouralEventWrapper = behaviouralEventWrapper,
-        )
+        Column {
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/appsearchmodal/BpkAppSearchModal.kt:99]: why this change?

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/appsearchmodal/BpkAppSearchModal.kt [hunk 5]
```diff
@@ -70,8 +72,7 @@ data class BpkAction(val text: String, val onActionSelected: () -> Unit)
 @Composable
 fun BpkAppSearchModal(
     title: String,
-    inputText: String,
-    inputHint: String,
+    inputSummary: SearchInputSummary,
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/appsearchmodal/BpkAppSearchModal.kt:75]: rather than making the breaking change please add a `prefix` parameter (like the one in the `SearchInputSummary`) and set the default to `icon`

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/searchinputsummary/BpkSearchInputSummary.kt [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import net.skyscanner.backpack.compose.icon.BpkIcon
+import net.skyscanner.backpack.compose.icon.BpkIconSize
+import net.skyscanner.backpack.compose.text.BpkText
+import net.skyscanner.backpack.compose.textfield.BpkClearAction
+import net.skyscanner.backpack.compose.textfield.BpkTextField
+import net.skyscanner.backpack.compose.theme.BpkTheme
+import net.skyscanner.backpack.compose.tokens.BpkSpacing
+import net.skyscanner.backpack.compose.tokens.Search
+
+data class SearchInputSummary(
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/searchinputsummary/BpkSearchInputSummary.kt:36]: we generally don't have data classes for this and prefer a flat list of params. can you please remove this and set the params directly, only keeping the sealed class for the prefix

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/textfield/BpkTextField.kt [hunk 7]
```diff
@@ -58,7 +63,55 @@ fun BpkTextField(
...
+        keyboardOptions = keyboardOptions,
+        keyboardActions = keyboardActions,
+        maxLines = maxLines,
+        interactionSource = interactionSource,
+        clearAction = clearAction,
+    )
+}
+
+@Composable
+fun BpkTextField(
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/textfield/BpkTextField.kt:93]: can you make this internal please? this should not be exposed to consumers (or just use the impl function from the search input summary)

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/textfield/BpkTextField.kt (hunk 2 above)
Comment (marianeum) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/textfield/BpkTextField.kt:66]: maybe the internal text field implementation can change to take the prefix sealed class from the search input summary, so we only need to decide in one place how to render an icon or text and keep it consistent

File: docs/compose/AppSearchModal/README.md [hunk 8]
```diff
@@ -40,13 +46,15 @@ BpkAppSearchModal(
     inputText = destination.value,
     inputHint = stringResource(id = R.string.text_field_hint),
     results = BpkAppSearchModalResult.Content(
-        sections = listOf(BpkSection(...)),
-        shortcuts = listOf(BpkShortcut(...)),
+        sections = listOf(BpkSection(...)
```
Comment (marianeum) [inline docs/compose/AppSearchModal/README.md:49]: looks like the formatting broke in the readme - can you please fix that? it's quite hard to parse right now

**Review by marianeum — COMMENTED** [review]
nice one - mainly some comments around the API

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/appsearchmodal/BpkAppSearchModal.kt (hunk 4 above)
Comment (henrik-sky) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/appsearchmodal/BpkAppSearchModal.kt:99]: A remnant from a previous way of implementing the change. Will remove

File: docs/compose/AppSearchModal/README.md (hunk 8 above)
Comment (marianeum) [reply docs/compose/AppSearchModal/README.md:49]: this still looks funky - i think the auto formatter gets confused about the `...` - sections + shortcuts are at the same level, but it looks like shortcuts is a level up because of the formatting - have a look at the diff here to make it clearer what i mean

File: docs/compose/AppSearchModal/README.md [hunk 9]
```diff
@@ -77,13 +83,39 @@ import net.skyscanner.backpack.compose.textfield.BpkClearAction
 
 val destination = remember { mutableStateOf(inputText) }
 BpkAppSearchModal(
+    modifier = modifier,
     title = stringResource(id = R.string.destination),
     inputText = destination.value,
     inputHint = stringResource(id = R.string.text_field_hint),
-    results = BpkAppSearchModalResult.Error(...),
-    closeAccessibilityLabel = stringResource(id = R.string.navigation_close),
+    results = BpkAppSearchModalResult.Error(...
```
Comment (marianeum) [inline docs/compose/AppSearchModal/README.md:90]: same funkyness down here + from line 92 there's no indentation at all

File: docs/compose/AppSearchModal/README.md [hunk 10]
```diff
@@ -77,13 +83,39 @@ import net.skyscanner.backpack.compose.textfield.BpkClearAction
...
+import net.skyscanner.backpack.compose.appsearchmodal.BpkAppSearchModal
+import net.skyscanner.backpack.compose.textfield.BpkClearAction
+
+val destination = remember { mutableStateOf(inputText) }
+BpkAppSearchModal(
+    title = stringResource(id = R.string.destination),
+    inputText = destination.value,
+    inputHint = stringResource(id = R.string.text_field_hint),
+    results = BpkAppSearchModalResult.Content(
+        sections = listOf(BpkSection(...)
```
Comment (marianeum) [inline docs/compose/AppSearchModal/README.md:112]: and same section/shortcuts issue here

**Review by marianeum — COMMENTED** [review]
looks good - noticed a couple inconsistencies with figma:
- the border radius should be 12 and the spacing for the close icon on the right 16 (these are both different to the text input which should stay as it is) probably best to add a `type` enum for text field vs search input summary to the text field implementation api and switching based on that. not ideal, but as both share a lot of functionality + styles that's probably the cleanest approach

File: docs/compose/AppSearchModal/README.md (hunk 10 above)
Comment (henrik-sky) [reply docs/compose/AppSearchModal/README.md:112]: There's a missing bracket I've just spotted 🙃

Comment (marianeum) [reply docs/compose/AppSearchModal/README.md:112]: no you can see the closing bracket is still on the next line - both sections + shortcuts are part of `content` but the formatting makes it look like shortcuts is outside of it

Comment (henrik-sky) [reply docs/compose/AppSearchModal/README.md:112]: It's the auto formatting, just because `...` isn't real Kotlin code 😮‍💨 must have fixed the same block 3 times over before realising.

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/icon/BpkIcon.kt [hunk 11]
```diff
@@ -51,6 +52,7 @@ class BpkIcon internal constructor(
 
 enum class BpkIconSize {
     Small,
+    Medium,
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/icon/BpkIcon.kt:55]: the icon is the large size (you always need to select the parent of the icon to ensure you get the icon size, rather than just that specific svg size without any border) - no need to add a new icon size

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/textfield/internal/BpkTextFieldImpl.kt [hunk 12]
```diff
@@ -315,15 +321,23 @@ private fun RowScope.TrailingIcon(
...
                 tint = icon.color,
-                modifier = icon.modifier,
+                modifier = icon.modifier.padding(if (type == BpkTextFieldType.SEARCH) BpkSpacing.Sm else 0.dp),
             )
         }
     }
 }
 
-private val Shape = RoundedCornerShape(BpkBorderRadius.Sm)
+internal enum class BpkTextFieldType {
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/textfield/internal/BpkTextFieldImpl.kt:332]: nit: we don't use all caps for enums

**Review by marianeum — COMMENTED** [review]
almost there

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/icon/BpkIcon.kt (hunk 11 above)
Comment (henrik-sky) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/icon/BpkIcon.kt:55]: Sorry, I don't understand what you're saying here

Comment (henrik-sky) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/icon/BpkIcon.kt:55]: I got it, on the Figma

**Review by marianeum — APPROVED** [review]
nice one! thanks for the contribution
